File: 202-config/Conversion/Ledger/Amount.php
<?php

declare(strict_types=1);

namespace Prosper202\Conversion\Ledger;

use InvalidArgumentException;

/**
 * Exact money arithmetic for the ledger, in integer units of 0.00001.
 *
 * Conversion and click amounts are `decimal(11,5)`. Summing them as floats
 * drifts (0.1 + 0.2 is not 0.3), and a click value derived from its rows
 * has to equal what a report re-adding the same rows gets, so the ledger
 * never adds floats: every amount is parsed once into an integer count of
 * the column's smallest unit and formatted back only when written.
 */
final class Amount
{
    public const SCALE = 5;
    private const FACTOR = 100000;

    private function __construct()
    {
    }

    /**
     * Parse a decimal amount into units.
     *
     * Accepts what MySQL returns for a decimal column and what a caller
     * passes as a payout: an int, a float, or a plain decimal string with an
     * optional leading minus. More than five decimal places is rounded half
     * away from zero, the way MySQL stores an over-precise value into
     * decimal(11,5). Anything else (an exponent, a thousands separator, an
     * empty string, a non-finite float) is refused: an amount that cannot be
     * read must never become 0 (CLAUDE.md error pattern #4).
     */
    public static function toUnits(int|float|string $value): int
    {
        if (is_int($value)) {
            return $value * self::FACTOR;
        }
        if (is_float($value)) {
            if (!is_finite($value)) {
                throw new InvalidArgumentException('amount is not a finite number');
            }
            // Round through the decimal string PHP would print, so 0.1 reads
            // as exactly 10000 units rather than 9999.999... truncated.
            $value = number_format($value, self::SCALE + 3, '.', '');
        }

        $text = trim($value);
        if (preg_match('/^(-)?(\d+)(?:\.(\d+))?$/D', $text, $m) !== 1) {
            throw new InvalidArgumentException('amount "' . $value . '" is not a decimal number');
        }
        $negative = $m[1] === '-';
        $whole = ltrim($m[2], '0');
        $fraction = $m[3] ?? '';
        if (strlen($whole) > 13) {
            throw new InvalidArgumentException('amount "' . $value . '" is too large');
        }

        $roundUp = false;
        if (strlen($fraction) > self::SCALE) {
            $roundUp = (int) $fraction[self::SCALE] >= 5;
            $fraction = substr($fraction, 0, self::SCALE);
        }
        $fraction = str_pad($fraction, self::SCALE, '0');

        $units = (int) ($whole === '' ? '0' : $whole) * self::FACTOR + (int) $fraction;
        if ($roundUp) {
            $units++;
        }

        return $negative ? -$units : $units;
    }

    /** Format units as the decimal string a decimal(11,5) column stores. */
    public static function fromUnits(int $units): string
    {
        $negative = $units < 0;
        $abs = abs($units);
        $text = intdiv($abs, self::FACTOR) . '.' . str_pad((string) ($abs % self::FACTOR), self::SCALE, '0', STR_PAD_LEFT);

        return ($negative ? '-' : '') . $text;
    }
}

File: sdk/ios-attribution/Sources/P202Attribution/Amount.swift
import Foundation

/// Money and sums in integer units of 0.00001, as the server's ledger keeps
/// them (`Prosper202\Conversion\Ledger\Amount`). The goal evaluator sums in
/// units so that `0.1 + 0.2` reaches a `gte 0.3` threshold on the device
/// exactly when it does on the server.
enum Amount {
    static let scale = 5
    static let factor = 100_000
    /// The ledger's decimal(11,5).
    static let maxUnits = 99_999_999_999

    /// A decimal string of digits with an optional minus and fraction, in
    /// units, rounding half away from zero past the fifth place. nil for
    /// anything else (an exponent, spaces inside, more than 13 whole digits).
    static func units(decimal text: String) -> Int? {
        var chars = Substring(trimPHP(text))
        var negative = false
        if chars.first == "-" {
            negative = true
            chars = chars.dropFirst()
        }
        let parts = chars.split(separator: ".", maxSplits: 1, omittingEmptySubsequences: false)
        guard let wholePart = parts.first, !wholePart.isEmpty, wholePart.allSatisfy(\.isASCIIDigit) else {
            return nil
        }
        var fraction = parts.count > 1 ? String(parts[1]) : ""
        if parts.count > 1 && (fraction.isEmpty || !fraction.allSatisfy(\.isASCIIDigit)) {
            return nil
        }
        let whole = String(wholePart.drop(while: { $0 == "0" }))
        guard whole.count <= 13 else {
            return nil
        }
        var roundUp = false
        if fraction.count > scale {
            let idx = fraction.index(fraction.startIndex, offsetBy: scale)
            roundUp = (fraction[idx].wholeNumberValue ?? 0) >= 5
            fraction = String(fraction[..<idx])
        }
        fraction += String(repeating: "0", count: scale - fraction.count)
        var units = (Int(whole.isEmpty ? "0" : whole) ?? 0) * factor + (Int(fraction) ?? 0)
        if roundUp {
            units += 1
        }
        return negative ? -units : units
    }

    /// An integer amount in units, or nil when it would overflow.
    static func units(int value: Int) -> Int? {
        let (product, overflow) = value.multipliedReportingOverflow(by: factor)
        return overflow ? nil : product
    }

    /// A double in units: formatted with eight decimals, then rounded half
    /// away from zero to five (the server's number_format() path). nil for a
    /// value that is not finite or is too large to hold.
    static func units(double value: Double) -> Int? {
        guard value.isFinite, abs(value) <= 99_999_999_999_999.0 else {
            return nil
        }
        return units(decimal: String(format: "%.8f", value))
    }

    /// A JSON number in units (an int exactly, a double by the rule above).
    static func units(number: JSONValue) -> Int? {
        switch number {
        case let .int(i): return units(int: i)
        case let .double(d): return units(double: d)
        default: return nil
        }
    }

    /// Units as the decimal string a decimal(11,5) column stores: "4.99000".
    static func format(_ units: Int) -> String {
        let negative = units < 0
        let magnitude = units.magnitude
        let whole = magnitude / UInt(factor)
        let fraction = String(magnitude % UInt(factor))
        let padded = String(repeating: "0", count: scale - fraction.count) + fraction
        return (negative ? "-" : "") + "\(whole).\(padded)"
    }

    /// Units as the shortest decimal string with at least two places, the
    /// way a canonical definition writes an amount: "20.00", "0.12345".
    static func formatShort(_ units: Int) -> String {
        let text = format(units)
        let pieces = text.split(separator: ".", maxSplits: 1)
        var fraction = String(pieces[1])
        while fraction.last == "0" { fraction.removeLast() }
        while fraction.count < 2 { fraction += "0" }
        return "\(pieces[0]).\(fraction)"
    }
}

/// PHP's trim(): only " \t\n\r\0\x0B", never other Unicode whitespace, so
/// the device trims exactly what the server trims.
func trimPHP(_ s: String) -> String {
    let set: Set<Unicode.Scalar> = [" ", "\t", "\n", "\r", "\0", "\u{0B}"]
    let scalars = Array(s.unicodeScalars)
    var start = 0
    var end = scalars.count
    while start < end, set.contains(scalars[start]) { start += 1 }
    while end > start, set.contains(scalars[end - 1]) { end -= 1 }
    var view = String.UnicodeScalarView()
    view.append(contentsOf: scalars[start..<end])
    return String(view)
}

extension Character {
    var isASCIIDigit: Bool {
        return ("0"..."9").contains(self) && isASCII
    }
}
